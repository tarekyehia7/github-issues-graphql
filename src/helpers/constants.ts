export const constants = {
    repository: 'facebook',
    projectName: 'react',
	searchTerm: 'title,body',
    issuesPerPage: 10,
    githubUrl: 'https://github.com'
};