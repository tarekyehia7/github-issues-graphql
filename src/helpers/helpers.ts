import formatDistanceToNowStrict from 'date-fns/formatDistanceToNowStrict';

export const getPagesNumber = (totalIssuesCount: number, issuesPerPage: number): number => {
	const totalPages = Math.ceil(totalIssuesCount / issuesPerPage);
	return !totalPages ? 1 : totalPages;
};

export const formatDate = (date: string) => {
	try {
		const myDate = new Date(date);
		const myFormattedDate = formatDistanceToNowStrict(myDate, { addSuffix: true });
		return myFormattedDate;
	} catch (e) {
		console.error(e);
		return '';
	}
};
