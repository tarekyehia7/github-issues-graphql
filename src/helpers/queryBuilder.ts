import { StatusEnum } from "../pages/Issues/Issues";

interface Query {
    repo: string;
    input: string;
    is: string;
    in: string;
    status: StatusEnum;
}

export const buildQuery = (query: Query) => {
    return Object.entries(query)
        .reduce((str, [p, val]) => `${str}${p}:${val} `, '')
        .replace('status:', '')
        .replace('input:', '')
        .trim();
};