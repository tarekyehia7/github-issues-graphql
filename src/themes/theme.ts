import baseStyled, { ThemedStyledInterface } from 'styled-components';

export const theme = {
    color: {
        orange: '#ff5f00',
        lightBlack: '#191919',
        lightGrey: '#e7e7e7',
        white: '#ffffff',
        black: '#000000',
    },
    screens: {
        tablet: '(max-width: 1048px)'
    }
};

export type Theme = typeof theme;
export const styled = baseStyled as ThemedStyledInterface<Theme>;