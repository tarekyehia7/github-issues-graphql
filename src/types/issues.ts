export enum StatusEnum {
	open = 'is:open',
	closed = 'is:closed',
}

export type QueryType = {
	repo: string;
	is: string;
	status: StatusEnum;
	in: string;
	input: string;
};
