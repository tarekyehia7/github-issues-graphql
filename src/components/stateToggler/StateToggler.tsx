import React from 'react';

import {
	ButtonsContainer,
	CorrectIconStyled,
	OpenIssueIconStyled,
	Button,
} from './StateToggler.styled';
import { StatusEnum } from '../../types/issues';

type StateTogglerProps = {
	status: StatusEnum;
	onStateClick: (status: StatusEnum) => void;
};

export const StateToggler = ({ status, onStateClick }: StateTogglerProps) => {
	return (
		<ButtonsContainer>
			<Button
				isSelected={status === StatusEnum.open}
				onClick={() => onStateClick(StatusEnum.open)}
			>
				<OpenIssueIconStyled isSelected={status === StatusEnum.open} />
				Open
			</Button>
			<Button
				isSelected={status === StatusEnum.closed}
				onClick={() => onStateClick(StatusEnum.closed)}
			>
				<CorrectIconStyled isSelected={status === StatusEnum.closed} />
				Closed
			</Button>
		</ButtonsContainer>
	);
};
