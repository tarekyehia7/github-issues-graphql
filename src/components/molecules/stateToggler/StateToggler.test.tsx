import React from 'react';
import { fireEvent, render } from '@testing-library/react';
import 'jest-styled-components';

import { StateToggler, StateTogglerProps } from './StateToggler';
import { PageWithTheme } from '../../../helpers/testing/helpers';
import { StatusEnum } from '../../../types/issues';
import { theme } from '../../../themes';

const mockOnClick = jest.fn();

const props: StateTogglerProps = {
	status: StatusEnum.open,
	onStateClick: mockOnClick,
};

const renderPage = (status?: StatusEnum) => {
	const { container, getByText } = render(
		<PageWithTheme>
			<StateToggler {...props} status={status ?? props.status} />
		</PageWithTheme>,
	);

	return {
		container,
		getByText,
	};
};

describe('<StateToggler />', () => {
	it('Should matches snapshot', async () => {
		const { container } = renderPage();
		expect(container).toMatchSnapshot();
	});

	it('Should trigger Open button on click', async () => {
		const { getByText } = renderPage();

		fireEvent.click(getByText('Open'));
		expect(mockOnClick.mock.calls.length).toBe(1);
		expect(mockOnClick).toHaveBeenCalledWith(StatusEnum.open);
	});

	it('Should trigger Closed button on click', async () => {
		const { getByText } = renderPage();

		fireEvent.click(getByText('Closed'));
		expect(mockOnClick.mock.calls.length).toBe(1);
		expect(mockOnClick).toHaveBeenCalledWith(StatusEnum.closed);
	});

	it('Should have correct color for Open button', async () => {
		const { getByText } = renderPage(StatusEnum.open);

		expect(getByText('Open')).toHaveStyleRule(`color`, theme.colors.lightBlack);
	});

	it('Should have correct color for Closed button', async () => {
		const { getByText } = renderPage(StatusEnum.open);

		expect(getByText('Closed')).toHaveStyleRule(`color`, theme.colors.darkGray);
	});
});
