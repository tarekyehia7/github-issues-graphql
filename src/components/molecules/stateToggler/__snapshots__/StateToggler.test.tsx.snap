// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<StateToggler /> Should matches snapshot 1`] = `
.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  padding: 1rem;
}

.c1 {
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  padding: 0.5rem;
  font-size: inherit;
  color: #000000;
  -webkit-text-decoration: none;
  text-decoration: none;
  white-space: nowrap;
  cursor: pointer;
  -webkit-user-select: none;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  background-color: transparent;
  border: 0;
  -webkit-appearance: none;
  -webkit-appearance: none;
  -moz-appearance: none;
  appearance: none;
  color: #24292f;
  font-weight: bold;
}

.c3 {
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  padding: 0.5rem;
  font-size: inherit;
  color: #000000;
  -webkit-text-decoration: none;
  text-decoration: none;
  white-space: nowrap;
  cursor: pointer;
  -webkit-user-select: none;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  background-color: transparent;
  border: 0;
  -webkit-appearance: none;
  -webkit-appearance: none;
  -moz-appearance: none;
  appearance: none;
  color: #57606a;
}

.c2 {
  fill: #24292f;
  margin-right: 0.5rem;
}

.c4 {
  fill: #57606a;
  margin-right: 0.5rem;
}

<div>
  <div
    class="c0"
  >
    <button
      class="c1"
    >
      <svg
        aria-hidden="true"
        class="c2"
        data-view-component="true"
        height="16"
        version="1.1"
        viewBox="0 0 16 16"
        width="16"
      >
        <path
          d="M8 9.5a1.5 1.5 0 100-3 1.5 1.5 0 000 3z"
        />
        <path
          d="M8 0a8 8 0 100 16A8 8 0 008 0zM1.5 8a6.5 6.5 0 1113 0 6.5 6.5 0 01-13 0z"
        />
      </svg>
      Open
    </button>
    <button
      class="c3"
    >
      <svg
        aria-hidden="true"
        class="c4"
        data-view-component="true"
        height="16"
        version="1.1"
        viewBox="0 0 16 16"
        width="16"
      >
        <path
          d="M13.78 4.22a.75.75 0 010 1.06l-7.25 7.25a.75.75 0 01-1.06 0L2.22 9.28a.75.75 0 011.06-1.06L6 10.94l6.72-6.72a.75.75 0 011.06 0z"
        />
      </svg>
      Closed
    </button>
  </div>
</div>
`;
